src/Compilers/CSharp/Portable/FlowAnalysis/AbstractFlowPass.cs

```diff
@@ -2452,6 +2451,35 @@ bool learnFromOperator(BoundBinaryOperator binary)
 
                 return false;
             }
+
+            // This function is only for the specific scenario where the LHS of a binary '=='/'!=' operator is a conditional access
+            // and the RHS is either a constant value or is of a non-nullable value type.
+            void visitRightOperandWithAnyTransferFunction(BoundExpression rightOperand, ref TLocalState leftStateWhenNotNull)
```

Comment:
This implementation is the same shape as `VisitCatchBlockWithAnyTransferFunction`.

https://github.com/dotnet/roslyn/blob/c12c1dd499a928d95114b55f06cd78c40552e39f/src/Compilers/CSharp/Portable/FlowAnalysis/AbstractFlowPass.cs#L1793-L1815

The theory here is that in some dataflow analyses such as DataFlowsOut, we may "flip" the behavior of definite assignment within the `/*bind*/` region, so that assigning a variable really behaves as if you "unassigned" the variable. Inside this function, we can take any "unassignments" which occurred in the RHS and propagate them to both `this.State` and `stateWhenNotNull`.

We do this by taking the "bottom" state in which all variables are assigned, and keeping it "on the side" in `NonMonotonicState` when we visit the right operand. After the visit, the only "unassigned" variables in the `NonMonotonicState` were those unassigned by the right operand, and we can simply `Join` those unassignments into the `stateWhenNotNull`.

See `CondAccess_Equals_DataFlowsOut_02` for a test case which is impacted by this change.